### ISSUE TITLE
refactor: ニューラルネットワークの構造を見直す

### ISSUE THREAD

**Issue by gckohaku** [gckohaku/cnn-learn-rust#20]
ニューラルネットワークの構造を見直して開発しやすいようにしたい。
これを利用して、RNN などの実装も行いやすくなるのを期待したい。

##  構造

- ~~`Layer` というトレイトを用意する
現時点では `forward`, `backward` 関数の実装を強制する~~
- ~~`ConvolutionLayer`, `PoolingLayer`, `FullyConnectedLayer` 構造体に `Layer` トレイトを適用する~~
- ~~その他、共通化できそうな関数があれば、`Layer` トレイトの定義に加える~~
- ~~順伝播、逆伝播の両方で単一のモジュールであると捉えられる範囲~~ 順伝播時に分けられる最小単位 (？) をひとつのモジュールとし、その単位に `NNModule` トレイトを適用する
例えば、全結合層はアフィン変換部分と活性化関数部分に分かれるが、それらを `Linear`, `ReLU` (活性化関数が ReLU の場合) 構造体とし、それぞれ `NNModule` トレイトを適用する
- 名前空間は `nn_modules` として、`nn_modules::<モジュール名>` で呼び出せるようにする
例：`nn_modules::Linear`
- Pytorch の構造を参考にしてはいるが、完全に Pytorch と同じ構造にする必要はない

### `NNModule` トレイトを適用する構造体
- `nn_modules::Linear`
- `nn_modules::Convolution`
- `nn_modules::Pooling`
- `nn_modules::ReLU`
- `nn_modules::Softmax`
- `nn_modules::CrossEntropyLoss`
- `nn_modules::SoftmaxAndCELoss`
  ※ pytorch の `torch.nn.CrossEntropyLoss` に似たようなもの (pytorch では LogSoftmax らしい)
- `nn_modules::Sigmoid` (優先度低)
- `nn_modules::Arctan` (優先度低)
- `nn_modules::BinaryCELoss` (優先度低)
- `nn_modules::MSELoss` (優先度低)

## Builder の用意
すべての `NNModule` トレイトを適用している構造体は、Builder を用いて生成するようにする。
これは、オプションパラメータを持たない構造体であっても例外ではない。
この Builder に適用するトレイトを `NNModuleBuilder` とし、`new` メソッドと `build` メソッドを実装することを要請する。

## 勾配の求め方
ニューラルネットワークにおいて、勾配を求めることは誤差逆伝播法の実装に不可欠である。
順伝播を行うタイミングで勾配を求めておき、逆伝播時に利用できるようにする

**Comment by gckohaku** [comment]
`NNModule` の引数は ndarray の `ArrayViewD` にした方がいいかも
これで参照を持った関連型も必要なくなり、dyn を使った実装もできるようになる
戻り値はこれまで通り関連型を使った実装でいいかも

**Comment by gckohaku** [comment]
FCNN の部分では完了したが、CNN での実装がまだである
この issue は、CNN の実装も含めて完了したときに close とする

**Comment by gckohaku** [comment]
かなり時間がかかってしまったが、ニューラルネットワークの構造を見直し、再構成するという目標は達成できた
よって、この issue を close とする
完了にあたって、`refactor-remake-structure` ブランチから `develop` ブランチへの PR を作成する
